### ISSUE TITLE
Always Surprised

### ISSUE THREAD

**Issue by rkrishnan2012** [luca-m/emotime#38]
Hello!

I cloned your project and ported it to Android, so it now takes images from the camera and runs analysis on them to see expressions. However, I keep getting "Surprised" and Score=1.0. Could you tell me what size I need for the image (I am currently doing 48x48) and whether or not the image needs to be monochrome before I pass it to the FacePreProcessor?

Thanks,
Rohit

**Comment by dbellavista** [comment]
Hi! Thanks for the interest.

First of all, the image needs to be in greyscale, but the face detector should automatically convert the image.
Second the image size depends on how the training was performed, parameters must be equals.

But before that, did it work on your PC with the webcam?

**Comment by rkrishnan2012** [comment]
To be honest, it wasn't too great but I am thrilled that you guys got this far! I didn't actually train the data, and just used the sample data set you gave. Could you please guide me on how to make my own training files? 

I also realized that the names of the files cannot be modified since your code delimits them by underscores.

**Comment by dbellavista** [comment]
This was an academic project, once passed the exam we didn't have much time to continue the development, so there are many limitation :) one big problem is the face analysis in bad light conditions. We obtained nice results only with the right light, without glasses, etc.. And then the training parameters should be studied better to improve the results.

Just to say that any contributions is welcomed!

To do your own training (really time consuming) you need access to the cohn kanade database, wich it's given only for academic purposes. Once you have it, you can use the python script reported on the Readme.

**Comment by luca-m** [comment]
Hi rkrishnan2012,

Just few details: the size of the images of the faces used for training was around 170x170, this could be a reference size for your test. 
Actually all the machinery works as follow: face is detected, gabor features are calculated,and then resized to the size specified in config file (note for me: maybe we can change this order), prediction with svm.

However, as told by dbellavista,  feel free to contribute.. Any advice is always welcome!

PS: in the readme you can also find the link to the [CK dataset](http://www.consortium.ri.cmu.edu/ckagree/)

**Comment by rkrishnan2012** [comment]
Hey! I just want to get back to say that it works well not on android when I set the camera capture to monochrome and the size to 640x480. Also, currently a single frame takes 45 seconds to run the following lines of code. Would you know if I can make this any more efficient on the phone?

bool FacePreProcessor::filterImage(Mat& src, Mat& out) {
    Mat filtered;
    filtered = this->gaborbank.filterImage(src, this->imgsize);
    filtered.copyTo(out);
    //filtered.release();
    return true;
  }

**Comment by luca-m** [comment]
You got the point. The bottleneck is the application of the filtering bank at `src/gaborbank/GaborBank.cpp:224`  :

```
    Mat GaborBank::filterImage(cv::Mat & src, cv::Size & featSize) {
      unsigned int i;
      unsigned int j;
      unsigned int k;
      Size size(0,0);
      Size s = this->getFilteredImgSize(featSize);
      Mat dest = Mat(s, CV_32F);
      Mat image;
      size.height=src.rows;
      size.width=src.cols;
      if (size.height==0 || size.width==0){
        std::cerr<<"WARN: cannot filter image (size="<<size<<")"<<std::endl;
        return Mat();
      }
      src.convertTo(image, CV_32F);
      for (k = 0; k < bank.size(); k++) {
        emotime::GaborKernel * gk = bank.at(k);
        Mat real = gk->getReal();
        Mat imag = gk->getImag();
        Mat freal = Mat(size, CV_32F);
        Mat fimag = Mat(size, CV_32F);
        Mat magn  = Mat(size, CV_32F);
        Mat scaled  = Mat(featSize,CV_32F);
        filter2D(src, freal, CV_32F, real);
        filter2D(src, fimag, CV_32F, imag);
        // Calculating Gabor magnitude
        pow(freal,2,freal);
        pow(fimag,2,fimag);
        add(fimag,freal,magn);
        sqrt(magn,magn);
        resize(magn, scaled, featSize, 0, 0, CV_INTER_AREA);
...
```

In my opinion for a good improvement we could perform the resize before the application of the gabor filter bank, this way the filter will be no more applied on a 200x200 or more grayscal image but instead in a 96x96 or less image (depends on configs). The downside is that this modification could require a tuning of the gabor filter bank's paramenter in order to extract meaningful feature for the svm classification. I'll play a bit with it.

**Comment by rkrishnan2012** [comment]
Ok awesome, I am also going to try to get the filter2d to be done with TBB so it will run in parallel threads for the android:

http://www.jayrambhia.com/blog/opencv-with-tbb/

**Comment by luca-m** [comment]
Take a look at new commits, I think we solved the performance issue..

**Comment by luca-m** [comment]
performance issue solved with 95c2eb0b58972058ba31726e9b7f80d949d3be5f
